chore: Authenticating api server against webhook (#2359)
open-policy-agent/gatekeeper: website/docs/externaldata.md

```

2. If `cert-controller` is disabled via the `--disable-cert-rotation` flag, you can use a cluster-wide, well-known CA certificate for Gatekeeper so that your external data provider can trust it without being deployed to the `gatekeeper-system` namespace.

### Authenticate API server against Webhook (Self managed kube cluster only)

To have requests coming from apiserver authenticated by webhook, following configuration can be made:

> 1. Deploy gatekeeper with cert rotator disabled and with a client ca name. Provide name of the client ca with the flag `--client-cert-name`. The same name will be used to read certificate from the webhook secret. The webhook will only authenticate apiserver requests if both - cert rotator disabled and client ca name is provided with flag - of the criteria are met.

Reviewer: Is there a reason cert rotation needs to be disabled for this to work? Is there a way to make this orthogonal? Maybe store the cert in a separate secret/configmap (IIUI  WRT secret vs. configmap the cert is not sensitive)

Author: Interesting. Why can't we use the same secret? does cert rotator delete and creates new secret while rotating cert or does it just update the whole data field making it impossible to persist manually added entries to data map?

Reviewer: TBH I can't remember how cert-rotator behaves, but I wouldn't be surprised if it clobbered the whole data field.

In any case, if we could use the same secret I'd still like to enable/disable either option separately for operational flexibility.

```

2. If `cert-controller` is disabled via the `--disable-cert-rotation` flag, you can use a cluster-wide, well-known CA certificate for Gatekeeper so that your external data provider can trust it without being deployed to the `gatekeeper-system` namespace.

### Authenticate API server against Webhook (Self managed kube cluster only)

To have requests coming from apiserver authenticated by webhook, following configuration can be made:

> 1. Deploy gatekeeper a client ca name. Provide name of the client ca with the flag `--client-cert-name`. The same name will be used to read certificate from the webhook secret. The webhook will only authenticate apiserver requests if both client ca name is provided with flag.

Reviewer: nit "ca" -> "CA"

2. You will need to patch the webhook secret manually to attach client ca crt. Update secret `gatekeeper-webhook-server-cert` to include `clientca.crt`. Key name `clientca.crt` should match the name passed with `--client-cert-name` flag.

    ```
    kind: Secret
    apiVersion: v1
    data:
      ca.crt: <ca-crt>
      ca.key: <ca-key>
      tls.crt: <tls-crt>
      tls.key: <tls-key>
      clientca.crt: <apiserver's ca.crt>
    metadata:
      ...
      name: gatekeeper-webhook-server-cert
      namespace: gatekeeper-system
    type: Opaque
    ```

> 3. You will need to make sure that apiserver includes appropriate certificate while sending requests to apiserver, otherwise webhook will not accept these requests and will log error of `tls client didn't provide a certificate`. To make sure apiserver attaches correct certificate to requests being sent to webhook, you must specify the location of the admission control configuration file via the `--admission-control-config-file` flag while starting apiserver. Here is an example admission control configuration file

Reviewer: nit: add a colon (`:`) at the end of the line

    ```
    apiVersion: apiserver.config.k8s.io/v1
    kind: AdmissionConfiguration
    plugins:
    - name: ValidatingAdmissionWebhook
      configuration:
        apiVersion: apiserver.config.k8s.io/v1
        kind: WebhookAdmissionConfiguration
        kubeConfigFile: "<path-to-kubeconfig-file>"
    - name: MutatingAdmissionWebhook
      configuration:
        apiVersion: apiserver.config.k8s.io/v1
        kind: WebhookAdmissionConfiguration
        kubeConfigFile: "<path-to-kubeconfig-file>"
    ```
  
    KubeConfig file should look something like:

    ```
    apiVersion: v1
    clusters:
    - cluster:
        certificate-authority-data: <clientca.crt> # same value as provided in gatekeeper webhook secret's clientca.crt 

>         server: https://gatekeeper-webhook-service.gatekeeper-system.svc:443

Reviewer 2: can we add note about service or namespace name and it should be updated if it's different

### Authenticate API server against Webhook (Self managed kube cluster only)

To have requests coming from apiserver authenticated by webhook, following configuration can be made:

1. Deploy gatekeeper a client CA name. Provide name of the client ca with the flag `--client-cert-name`. The same name will be used to read certificate from the webhook secret. The webhook will only authenticate apiserver requests if both client ca name is provided with flag.

2. You will need to patch the webhook secret manually to attach client ca crt. Update secret `gatekeeper-webhook-server-cert` to include `clientca.crt`. Key name `clientca.crt` should match the name passed with `--client-cert-name` flag.

    ```
    kind: Secret
    apiVersion: v1
    data:
      ca.crt: <ca-crt>
      ca.key: <ca-key>
      tls.crt: <tls-crt>
      tls.key: <tls-key>
      clientca.crt: <apiserver's ca.crt>
    metadata:
      ...
      name: gatekeeper-webhook-server-cert

>       namespace: gatekeeper-system

Reviewer 2: let's update here too re: name and namespace

```

2. If `cert-controller` is disabled via the `--disable-cert-rotation` flag, you can use a cluster-wide, well-known CA certificate for Gatekeeper so that your external data provider can trust it without being deployed to the `gatekeeper-system` namespace.

### Authenticate API server against Webhook (Self managed kube cluster only)

> To have requests coming from apiserver authenticated by webhook, following configuration can be made:

Reviewer 3: ```suggestion
To ensure a request to the Gatekeeper webhook is coming from the api server, Gatekeeper needs to validate the client cert in the request. To enable authenticate api server, the following configuration can be made:
```

> 1. Deploy gatekeeper a client CA name. Provide name of the client ca with the flag `--client-cert-name`. The same name will be used to read certificate from the webhook secret. The webhook will only authenticate apiserver requests if both client ca name is provided with flag.

Reviewer 3: ```suggestion
1. Deploy gatekeeper with a client CA cert name. Provide name of the client CA with the flag `--client-cert-name`. The same name will be used to read certificate from the webhook secret. The webhook will only authenticate apiserver requests if both client ca name is provided with flag.
```

Reviewer 3: > if both client CA name is provided with flag

what do you mean by "both" here?

Author: Ohh! "both" isn't supposed to be here.

2. You will need to patch the webhook secret manually to attach client ca crt. Update secret `gatekeeper-webhook-server-cert` to include `clientca.crt`. Key name `clientca.crt` should match the name passed with `--client-cert-name` flag.

    ```
    kind: Secret
    apiVersion: v1
    data:
      ca.crt: <ca-crt>
      ca.key: <ca-key>
      tls.crt: <tls-crt>
      tls.key: <tls-key>
      clientca.crt: <apiserver's ca.crt>
    metadata:
      ...
      name: <gatekeeper-webhook-service-name>
      namespace: <gatekeeper-namespace>
    type: Opaque
    ```

> 3. You will need to make sure that apiserver includes appropriate certificate while sending requests to apiserver, otherwise webhook will not accept these requests and will log error of `tls client didn't provide a certificate`. To make sure apiserver attaches correct certificate to requests being sent to webhook, you must specify the location of the admission control configuration file via the `--admission-control-config-file` flag while starting apiserver. Here is an example admission control configuration file:

Reviewer 3: ```suggestion
3. You will need to make sure that apiserver includes appropriate certificate while sending requests to the webhook, otherwise webhook will not accept these requests and will log error of `tls client didn't provide a certificate`. To make sure apiserver attaches correct certificate to requests being sent to webhook, you must specify the location of the admission control configuration file via the `--admission-control-config-file` flag while starting apiserver. Here is an example admission control configuration file:
```
